### PR TITLE
feat: cross-val-score and k-fold implemented

### PR DESCRIPTION
**Pull request by DirkToewe**
Cross validation is not just crucial to machine learning itself. It is also useful to validate machine learning model implementations. This PR aims to add support for basic k-fold cross validation.

### CODE REVIEW

**Comment by risenW** [comment]
@DirkToewe is this ready for review? I need some these class in order to implement GridSearch

**Comment by DirkToewe** [comment]
It is ... I hope :smile:

File: src/model_selection/crossValScore.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { Scalar, Tensor1D, Tensor2D } from '@tensorflow/tfjs-node'
```
Comment (risenW) [inline src/model_selection/crossValScore.ts:16]: I think you should import types from the Global tf as well.

File: src/model_selection/kFold.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { tf } from '../shared/globals'
+import { Tensor2D, Tensor1D } from '@tensorflow/tfjs-node'
```
Comment (risenW) [inline src/model_selection/kFold.ts:17]: Same here. You should import the types from the global `tf`

File: src/model_selection/crossValidator.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { Tensor1D, Tensor2D } from '@tensorflow/tfjs-node'
```
Comment (risenW) [inline src/model_selection/crossValidator.ts:16]: Same here. Import types from global tf, so browser version don't break

File: src/model_selection/kFold.test.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+* ==========================================================================
+*/
+
+import { tf } from '../shared/globals'
+import * as fc from 'fast-check'
+import { KFold } from './kFold'
+import { alea } from 'seedrandom'
+import { Tensor2D } from '@tensorflow/tfjs'
```
Comment (risenW) [inline src/model_selection/kFold.test.ts:20]: Same here: Import type from Global tf

File: src/model_selection/kFold.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      nSplits % 1 === 0 && nSplits > 1,
+      'new KFold({nSplits}): nSplits must be an int greater than 1.'
+    )
+    this.nSplits = nSplits
+    this.shuffle = !!shuffle
+    this.randomState = randomState
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public getNumSplits(X?: Tensor2D, y?: Tensor1D, groups?: Tensor1D): number {
```
Comment (risenW) [inline src/model_selection/kFold.ts:100]: I have limited knowledge here, but I see you passed params but didn't use them

File: src/model_selection/kFold.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    return this.nSplits
+  }
+
+  public *split(
```
Comment (risenW) [inline src/model_selection/kFold.ts:104]: A suggestion: shouldn't we support other data types and not just Tensors? I believe the Scikit2D and Scikit1D types covers those

File: src/model_selection/kFold.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    if (shuffle) {
+      const rng = alea(randomState?.toString())
+      randUtils.shuffle(rng)(range)
+    }
+
+    const chunkBase = Math.floor(nSamples / nSplits)
+    let remainder = nSamples % nSplits
+
+    for (let offset = 0; offset < nSamples; ) {
+      const chunk = chunkBase + Number(remainder-- > 0)
```
Comment (risenW) [inline src/model_selection/kFold.ts:145]: Not sure what this is. Valid JS syntax?

File: src/model_selection/crossValScore.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    cv?: CrossValidator
+    groups?: Tensor1D
+    scoring: (this: T, X: Tensor2D, y: Tensor1D) => Scalar
+  }
+): Promise<Tensor1D>
+
+/**
+ * Evaluates a score by cross-validation. This particular overload
+ * of the function uses the given scorer to cross validate an
+ * unsuperwised estimator.
```
Comment (risenW) [inline src/model_selection/crossValScore.ts:62]:
```suggestion
 * unsupervised estimator.
```

File: src/model_selection/crossValScore.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+ *
+ * @returns        The test score for each split/fold that was generated by
+ *                 the {@link CrossValidator} `cv`
+ */
+export async function crossValScore(
+  estimator: {
+    fit(X: Tensor2D, y: Tensor1D): Promise<unknown>
+    score(X: Tensor2D, y: Tensor1D): Scalar
+  },
+  Xy: [Tensor2D, Tensor1D],
```
Comment (risenW) [inline src/model_selection/crossValScore.ts:122]: Better to separate these params?

**Review by risenW — CHANGES_REQUESTED** [review]
@DirkToewe Great work! I dropped a couple of comments and suggestions. @DirkToewe may have some thoughts as well. So pinging him.

File: src/model_selection/crossValScore.ts (hunk 1 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts:16]: That's I will fix that

File: src/model_selection/crossValScore.ts (hunk 9 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts:122]: The difficulty is that this method is supposed to work for unsupervised and supervised models. With supervised models the call would be:
```js
crossValScore(supervisedEstimator, X, y, {...})
```
And in the supervised case it would be:
```js
crossValScore(unsupervisedEstimator, X, {...})
```
But since the parameters are intended to be optional, the following call would also be valid:
```js
crossValScore(supervisedEstimator, X, y)
```
How to then decide whether y is a `Scikit1D` or a set of parameters? That is why I packed `[X, y]` into an array... but maybe there is a better way?

File: src/model_selection/kFold.ts (hunk 5 above)
Comment (DirkToewe) [reply src/model_selection/kFold.ts:100]: I keep forgetting that one can just drop parameters in JS :smile:

File: src/model_selection/kFold.ts (hunk 7 above)
Comment (DirkToewe) [reply src/model_selection/kFold.ts:145]: Yeah it is :smile: `Number(true) === 1` and `Number(false) === 0`. Basically the first `remainder` splits/chunks are one element larger than `chunkBase` to match sample size which are not evenly divisible by `nSplits`. It's equivalent to:
```js
const chunk = remainder-- > 0 ? chunkBase + 1 : chunkBase
```
Would that be more readable? I could change it.

**Comment by DirkToewe** [comment]
Most issues should be resolved now, except for the `[X, y]` issue... I thing I'm gonna sleep on that one. Maybe I can come up with a better solution. Suggestions are welcome.

File: src/model_selection/scorers.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+export function negMeanSquaredError(
+  this: {
+    predict(X: Tensor2D): Tensor1D
+  },
+  X: Scikit2D,
+  y: Scikit1D
+) {
+  X = convertToTensor2D(X)
+  y = convertToTensor1D(y)
+  const yPred = this.predict(X)
```
Comment (dcrescim) [inline src/model_selection/scorers.ts:45]: Usually the `Estimator` itself deals with `Scikit1D` and `Scikit2D`, so I don't think you need to explicitly make the Tensors here. I'm sure some estimators operate on Tensors under the hood, but some don't, and will end up converting them back to arrays (DecisionTree). So anyway, I think you can just pass them in wholesale.

No need for the  `X = convertToTensor2D(X)` above

File: src/neighbors/kNeighborsRegressor.test.ts [hunk 11]
```diff
@@ -13,65 +13,17 @@
...
-  if (len !== y.shape[0]) {
-    throw new Error('kFold(k,X,y): X.shape[0] must equal y.shape[0].')
-  }
-
-  for (const [train, test] of kFoldIndices(k, len)) {
-    yield [X.gather(train), y.gather(train), X.gather(test), y.gather(test)]
-  }
-}
+import { crossValScore } from '../model_selection/crossValScore'
+import { KFold } from '../model_selection/kFold'
```
Comment (dcrescim) [inline src/neighbors/kNeighborsRegressor.test.ts:21]: I love how squeaky clean this file got 👍 Good work on this whole PR @DirkToewe

File: src/model_selection/crossValScore.ts [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    cv = new KFold(),
+    groups,
+    scoring
+  }: {
+    cv?: CrossValidator
+    groups?: Scikit1D
+    scoring?: any
+  } = {}
+): Promise<Tensor1D> {
+  if (undefined === scoring) {
```
Comment (dcrescim) [inline src/model_selection/crossValScore.ts:179]: I always prefer `scoring === undefined` (instead of the reverse) cause it reads like a sentence in my brain. "If scoring is undefined then blah blah". This is personal preference and not a huge deal.

File: src/model_selection/crossValScore.ts [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+      } finally {
+        tf.engine().endScope(score)
+        trainIndex.dispose()
+        testIndex.dispose()
+      }
+    }
+
+    return (result = tf.stack(scores) as Tensor1D)
+  } finally {
+    tf.engine().endScope(result)
```
Comment (dcrescim) [inline src/model_selection/crossValScore.ts:217]: Same for the outer `try`. I'm just basically calling dispose on the Tensors to avoid the `try / finally / tf.engine` block. Here's the full code.

```
  X = convertToTensor2D(X)
  if (y != null) {
    y = convertToTensor1D(y)
  }

  for (const { trainIndex, testIndex } of cv.split(X, y, groups)) {
    let XTrain = X.gather(trainIndex)
    let yTrain = y?.gather(trainIndex)
    let XTest = X.gather(testIndex)
    let yTest = y?.gather(testIndex)

    await estimator.fit(XTrain, yTrain)
    let score = scoring(XTest, yTest) as Scalar

    scores.push(score)

    let cleanup = [XTrain, yTrain, XTest, yTest, trainIndex, testIndex]
    cleanup.forEach((el) => el?.dispose())
  }

  let cleanup = [X, y]
  cleanup.forEach((el) => el?.dispose())
  return tf.stack(scores) as Tensor1D
}
```

Comment (dcrescim) [reply src/model_selection/crossValScore.ts:217]: Also you'll notice that I just create this intermediate arrays and loop through them just optionally calling dispose. We could just create a helper function in `utils` called `cleanup` that does just that, and use it here, which would save a couple more lines.

Basically,
```
function cleanup(tensors: any[]){
  tensors.forEach(el => el?.dispose())
}
```

And then the above would just be
`cleanup([X, y])` or `cleanup([XTrain, yTrain, blah, blah])`

**Review by dcrescim — COMMENTED** [review]
@DirkToewe this is amazing work. I love the attention to detail 😄 
I put my notes into the PR mostly about cleaning this or that, but I wanted to talk through the `crossValScore` api here. You mention above that there are really two cases that we want to support. 
The unsupervised, supervised call to `crossValScore`.

So the two basic calls are just

`crossValScore(clf, X, y, params)` and `crossValScore(clf, X, params)`

Why can't we just look at the third argument to decide if they passed a Scikit1D or not? Maybe this is more complicated than I'm understanding but we just have to check that `isScikit1D(y)` is true or not, right? That func exists in the utils, I think. And if that function call returns false, then we assume that it is a parameter object. I call this solution option 1 below.

I'm not immune to the downsides of this approach which are basically the Typescript type is strange for that variable. I think it would be like
clf -> `Estimator`
X -> `Scikit2D`
y -> optional `Scikit1D` | `paramObject`
params -> optional `paramObject`

I still think that is better than packing them into a `[X, y]` array because there is no precedent for that on the sklearn side. 

Another approach would be to "mandate" that the third argument is a Scikit1D, and just explicitly pass `undefined` there if we need to set the param object. I call this approach option 2. Function calls would like look

```
crossValScore(clf, X, undefined, { groups: ... })
```

Basically my vote is that we don't pack into an array, and either do option 1 or 2 above.

File: src/model_selection/crossValScore.ts (hunk 13 above)
Comment (steveoni) [reply src/model_selection/crossValScore.ts:217]: > There might be a way to get rid of the outer `try` / `finally` / `tf.engine`, by just manually calling dispose. I personally think it is easier to reason about, then nesting a double try here.
> 
> ```
>   X = convertToTensor2D(X)
>   if (y != null) {
>     y = convertToTensor1D(y)
>   }
> 
>   for (const { trainIndex, testIndex } of cv.split(X, y, groups)) {
>     let XTrain = X.gather(trainIndex)
>     let yTrain = y?.gather(trainIndex)
>     let XTest = X.gather(testIndex)
>     let yTest = y?.gather(testIndex)
> 
>     await estimator.fit(XTrain, yTrain)
>     let score = scoring(XTest, yTest) as Scalar
> 
>     scores.push(score)
> 
>     let cleanup = [XTrain, yTrain, XTest, yTest, trainIndex, testIndex]
>     cleanup.forEach((el) => el?.dispose())
>   }
> 
>   let cleanup = [X, y]
>   cleanup.forEach((el) => el?.dispose())
>   return tf.stack(scores) as Tensor1D
> }
> ```

@dcrescim I just want to point out the use of `tf.engine()` ; I was doing a project where I was having a memory leak. i tried out `tf.tidy` and `dispose` and yet there was still a bit of memory leak. but with `tf.engine()` there was no memory leak.

File: src/model_selection/kFold.ts (hunk 7 above)
Comment (risenW) [reply src/model_selection/kFold.ts:157]: Same here. Ternary is more readable 😀

File: src/model_selection/crossValScore.ts (hunk 9 above)
Comment (risenW) [reply src/model_selection/crossValScore.ts]: Oh, I see. I always go with an intuitive API since JavaScript does not support named parameters. That is all required params are positional and all optional params are part of the Param object. 

So in this case something like:

```javascript
crossValScore(unsupervisedEstimator | supervisedEstimator, X, { y, ...})
```

Another suggestion would be to make the second param an object, so the order does not matter. Example:
```javascript
 crossValScore(unsupervisedEstimator | supervisedEstimator, { X, y }, {...})
```

@DirkToewe @dcrescim  Thoughts?

**Comment by risenW** [comment]
> @DirkToewe this is amazing work. I love the attention to detail 😄 I put my notes into the PR mostly about cleaning this or that, but I wanted to talk through the `crossValScore` api here. You mention above that there are really two cases that we want to support. The unsupervised, supervised call to `crossValScore`.
> 
> So the two basic calls are just
> 
> `crossValScore(clf, X, y, params)` and `crossValScore(clf, X, params)`
> 
> Why can't we just look at the third argument to decide if they passed a Scikit1D or not? Maybe this is more complicated than I'm understanding but we just have to check that `isScikit1D(y)` is true or not, right? That func exists in the utils, I think. And if that function call returns false, then we assume that it is a parameter object. I call this solution option 1 below.
> 
> I'm not immune to the downsides of this approach which are basically the Typescript type is strange for that variable. I think it would be like clf -> `Estimator` X -> `Scikit2D` y -> optional `Scikit1D` | `paramObject` params -> optional `paramObject`
> 
> I still think that is better than packing them into a `[X, y]` array because there is no precedent for that on the sklearn side.
> 
> Another approach would be to "mandate" that the third argument is a Scikit1D, and just explicitly pass `undefined` there if we need to set the param object. I call this approach option 2. Function calls would like look
> 
> ```
> crossValScore(clf, X, undefined, { groups: ... })
> ```
> 
> Basically my vote is that we don't pack into an array, and either do option 1 or 2 above.

Possible solution to this here: https://github.com/javascriptdata/scikit.js/pull/186#discussion_r786446628

File: src/model_selection/crossValScore.ts (hunk 13 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts:217]: `try {} finally {}` is a language feature and therefore not really replaceable by anything. You could hide it behind some `tidyAsync`. But that would likely just be more verbose. Even if we switch from `beginScope()` and `endScope()` to dispose, `try {} finally {}` remains irreplaceable. Just look at the following code:

```js
X = convertToTensor2D(X)
y = convertToTensor1D(y)
X.dispose()
y.dispose()
```
If the second line `y = convertToTensor1D(y)` throws an `IllegalArgumentException` because the input is for some reason invalid, then `X` is never disposed, polluting memory for the rest of the runtime.

TFJS' explicit memory management is a nightmare, especially since JS in is not at all suited for anything like this. Using `beginScope()` and `endScope()` do bring the risk other async() code of some other function being executed in the middle of the scope of this function. Basically all calls would have to go through a `mutex`, which might be implementable in terms of a `tidyAsync` function. But it's still kind of nightmarish.

Calling `dispose()` explicitly, however, is also kind of bad:
  * Prevents the usage of `tf.keep`
  * Will not clean up any memory allocated by `estimator.fit`
  * Makes it very easy to forget one tensor or another

File: src/neighbors/kNeighborsRegressor.test.ts (hunk 11 above)
Comment (DirkToewe) [reply src/neighbors/kNeighborsRegressor.test.ts:21]: That was the motivation for taking on this PR :smile:

File: src/model_selection/crossValScore.ts (hunk 12 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts]: I tend to do it the other way around because it prevents accidents like `if(scoring = undefined)`. I changed it.

File: src/model_selection/crossValScore.ts (hunk 13 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts:238]: I've gotten rid of one try {} finally {} at least... It's still not ideal but I hope it will do until we have a more standard concept of how we want to handle memory.

File: src/model_selection/kFold.ts [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    } finally {
+      tf.engine().endScope()
+    }
+
+    const range = new Int32Array(nSamples)
+    for (let i = 0; i < range.length; i++) {
+      range[i] = i
+    }
+    if (shuffle) {
+      const rng = alea(randomState?.toString())
```
[the remark this answers is not in the document]
Comment (DirkToewe) [reply src/model_selection/kFold.ts:140]: We could move this `randUtils` as `createRng( seed?: String ): () => number`.

File: src/model_selection/crossValScore.ts (hunk 9 above)
Comment (DirkToewe) [reply src/model_selection/crossValScore.ts]: It should work as intended now. Both `crossValsScore(unsupervised, X, undefined, {})` and `crossValsScore(unsupervised, X, {})` should work.

File: src/model_selection/kFold.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (dcrescim) [reply src/model_selection/kFold.ts:140]: Love it. That's the play

File: src/model_selection/crossValScore.ts (hunk 13 above)
Comment (dcrescim) [reply src/model_selection/crossValScore.ts:238]: Works for me

File: src/model_selection/kFold.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  public *split(
+    X: Scikit2D,
+    y?: Scikit1D,
+    groups?: Scikit1D
+  ): IterableIterator<{ trainIndex: Tensor1D; testIndex: Tensor1D }> {
+    const { nSplits, shuffle, randomState } = this
+
+    let nSamples: number
+
+    nSamples = getLength(X)
```
Comment (dcrescim) [inline src/model_selection/kFold.ts:113]: You can replace the last two lines with 
`let nSamples = getLength(X)`

File: src/model_selection/kFold.ts [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+    assert(
+      nSplits <= nSamples,
+      'KFold({nSplits})::split(X): nSplits must not be greater than X.shape[0].'
+    )
+
+    if (null != y) {
```
Comment (dcrescim) [inline src/model_selection/kFold.ts:120]: `if (y != null)`

File: src/model_selection/kFold.ts [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+      assert(
+        nSamples === getLength(y),
+        'KFold::split(X,y): X.shape[0] must equal y.shape[0].'
+      )
+    }
+
+    if (null != groups) {
```
Comment (dcrescim) [inline src/model_selection/kFold.ts:127]: `if (groups != null)`

File: src/model_selection/scorers.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+  this: {
+    predict(X: Scikit2D): Tensor1D
+  },
+  X: Scikit2D,
+  y: Scikit1D
+): Scalar {
+  return tf.tidy(() => {
+    y = convertToTensor1D(y)
+    const yPred = this.predict(X)
+    return tf.equal(y, yPred).sum().div(y.shape[0]) as Scalar
```
Comment (dcrescim) [inline src/model_selection/scorers.ts:60]: Is there a reason why you aren't just using the functions in metrics? They are a thin wrapper around the tfjs api, and they already do the conversions to / from Scikit1D as well as some sanity checks on the input. Like this whole function could just be

```
export function accuracy(
  this: {
    predict(X: Scikit2D): Tensor1D
  },
  X: Scikit2D,
  y: Scikit1D
): Scalar {
  return tf.tidy(() => {
    return tf.scalar(metrics.accuracyScore(y, this.predict(X))) as Scalar
  })
}
```

Right now, all the functions in metrics return a `number` but if you want the result to be a `Scalar` instead of a `number`, we can simply change those functions to optionally return a Scalar. Like have a function call that looks like

`metrics.accuracyScore(labels, predictions) // returns number`
`metrics.accuracyScore(labels, predictions, true) // returns Scaler`

**Review by dcrescim — APPROVED** [review]
I added a couple notes. They are basically small cleanups (switch to ternary, switch if conditions), and possibility of using the `metrics` functions. Other than that, this is good to go. ✅  from me.

File: src/model_selection/scorers.ts (hunk 18 above)
Comment (DirkToewe) [reply src/model_selection/scorers.ts:60]: If this doesn't break anything sure. In my opinion `Scalar` is the datatype of choice because it allows to use sweet sweet backpropagation to optimize hyperparameters.

Once the changes are made to `metrics.ts`, we can even use a `makeScorer` as they do in `sklearn`.

But maybe that's a story for another commit? The commit is growing bigger and bigger touching more and more files.
